perf(semantic): reduce `AstNodeId` to `u32`

overlookmotel opened the pull request:
`AstNodeId` was a `usize`. This seems excessive.

Parser has a limit on size of a JS file of 4 GiB. While it is *possible* for a JS file of that size to create an AST with more than `1 << 32` (~4 billion) AST nodes, that would be insanely large.

So make `AstNodeId` `u32` instead.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4264?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4264** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4264?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4263** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4263?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#4262** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4262?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
*Now* we're talking!

<img width="737" alt="bench" src="https://github.com/user-attachments/assets/36cf1e3e-0ab8-4156-91a2-954685a26cd4">

overlookmotel commented:
@rzvxa When we were chatting a while back, you argued against making it a `u32`, in order to support very large files. However, I think it's hard to argue against this kind of performance boost. What do you think?

rzvxa commented:
> @rzvxa When we were chatting a while back, you argued against making it a `u32`, in order to support very large files. However, I think it's hard to argue against this kind of performance boost. What do you think?

Yep, We can add a feature later on for `large-file-support` that would use `usize`.

rzvxa reviewed (approved):
It LGTM; it is up to you @Boshen!

overlookmotel committed:
perf(semantic): reduce `AstNodeId` to `u32` (#4264)

`AstNodeId` was a `usize`. This seems excessive.

Parser has a limit on size of a JS file of 4 GiB. While it is *possible* for a JS file of that size to create an AST with more than `1 << 32` (~4 billion) AST nodes, that would be insanely large.

So make `AstNodeId` `u32` instead.
